File: src/kernel.js
const {
  findingCounts,
  invariant,
  isMaterialFinding,
  validateDecision,
  validateEvidence,
  validateFinding,
  validateProposal,
} = require("./domain");

function nowIso() {
  return new Date().toISOString();
}

function makeEvidenceId(session) {
  return `E${session.evidence.length + 1}`;
}

function evidenceIndex(session) {
  return new Map(session.evidence.map((item) => [item.id, item]));
}

function findingIndex(findings) {
  return new Map(findings.map((item) => [item.id, item]));
}

function currentRoundFindings(roundRecord) {
  return Array.isArray(roundRecord?.findings_against_proposal)
    ? roundRecord.findings_against_proposal
    : Array.isArray(roundRecord?.findings)
      ? roundRecord.findings
      : [];
}

function ensureOpenRound(session, round) {
  invariant(Number.isInteger(round) && round > 0, "open round index must be a positive integer");
  if (!session.open_round) {
    session.open_round = {
      index: round,
      proposal: null,
      evidence_added: [],
      findings_against_proposal: [],
      review_summary: { total: 0, high: 0, medium: 0, low: 0, material: 0, gaps: 0 },
      verdict: null,
      phase_history: [],
      created_at: nowIso(),
      updated_at: nowIso(),
      error: null,
    };
  }
  invariant(session.open_round.index === round, `session already has open round ${session.open_round.index}`);
  return session.open_round;
}

function refreshOpenRoundSummary(session) {
  const roundRecord = session.open_round;
  if (!roundRecord) {
    return;
  }
  const counts = findingCounts(currentRoundFindings(roundRecord));
  roundRecord.review_summary = counts;
  roundRecord.updated_at = nowIso();
  session.status = {
    ...session.status,
    round: roundRecord.index,
    unresolved_high: counts.high,
    unresolved_medium: counts.medium,
  };
}

function roundPhaseIndex(roundRecord, actor, phase, status = "running") {
  for (let index = roundRecord.phase_history.length - 1; index >= 0; index -= 1) {
    const item = roundRecord.phase_history[index];
    if (item.actor === actor && item.phase === phase && item.status === status) {
      return index;
    }
  }
  return -1;
}

function createSession({ id, topic, chair = "chair", critics = [], max_rounds = 3, adapter = "unknown" }) {
  invariant(typeof id === "string" && /^[A-Za-z0-9._-]+$/.test(id), "session id must match [A-Za-z0-9._-]+");
  invariant(typeof topic === "string" && topic.trim().length > 0, "topic must be a non-empty string");
  invariant(typeof chair === "string" && chair.trim().length > 0, "chair must be a non-empty string");
  invariant(Array.isArray(critics), "critics must be an array");
  invariant(Number.isInteger(max_rounds) && max_rounds > 0, "max_rounds must be a positive integer");
  invariant(typeof adapter === "string" && adapter.trim().length > 0, "adapter must be a non-empty string");

  return {
    version: 1,
    id,
    topic,
    created_at: nowIso(),
    chair,
    critics: [...critics],
    max_rounds,
    adapter,
    evidence: [],
    rounds: [],
    open_round: null,
    adjudicated_proposal: null,
    status: {
      round: 0,
      converged: false,
      unresolved_high: 0,
      unresolved_medium: 0,
      state: "initialized",
      active_actor: null,
      active_phase: null,
      error: null,
    },
  };
}

function appendEvidence(session, { items, collectedBy, phase, round }) {
  invariant(Array.isArray(items), "items must be an array");
  invariant(typeof collectedBy === "string" && collectedBy.trim().length > 0, "collectedBy must be a non-empty string");
  invariant(typeof phase === "string" && phase.trim().length > 0, "phase must be a non-empty string");
  invariant(Number.isInteger(round) && round >= 0, "round must be a non-negative integer");

  const added = [];
  for (const item of items) {
    invariant(item && typeof item === "object", "evidence input item must be an object");
    const evidence = {
      id: makeEvidenceId(session),
      source: item.source,
      kind: item.kind,
      phase,
      statement: item.statement,
      excerpt: item.excerpt,
      collected_by: collectedBy,
      round,
      created_at: nowIso(),
    };
    validateEvidence(evidence);
    session.evidence.push(evidence);
    added.push(evidence);
  }
  return added;
}

function startRound(session, round) {
  const roundRecord = ensureOpenRound(session, round);
  session.status = {
    ...session.status,
    round,
    converged: false,
    unresolved_high: roundRecord.review_summary.high || 0,
    unresolved_medium: roundRecord.review_summary.medium || 0,
    state: "running",
    active_actor: null,
    active_phase: null,
    error: null,
  };
  roundRecord.updated_at = nowIso();
  return roundRecord;
}

function recordPhaseStart(session, { round, actor, phase, input_summary }) {
  const roundRecord = startRound(session, round);
  roundRecord.phase_history.push({
    actor,
    phase,
    status: "running",
    input_summary: input_summary || null,
    output_summary: null,
    artifact: null,
    started_at: nowIso(),
    completed_at: null,
    duration_ms: null,
    error: null,
  });
  roundRecord.updated_at = nowIso();
  session.status = {
    ...session.status,
    active_actor: actor,
    active_phase: phase,
    error: null,
  };
  return roundRecord;
}

function completePhase(session, { round, actor, phase, output_summary, artifact = null, duration_ms, error = null }) {
  const roundRecord = ensureOpenRound(session, round);
  const index = roundPhaseIndex(roundRecord, actor, phase);
  invariant(index >= 0, `no running phase record for ${actor}/${phase} in round ${round}`);
  const nextStatus = error ? "failed" : "succeeded";
  const completedAt = nowIso();
  roundRecord.phase_history[index] = {
    ...roundRecord.phase_history[index],
    status: nextStatus,
    output_summary: output_summary || null,
    artifact,
    completed_at: completedAt,
    duration_ms: duration_ms ?? null,
    error: error || null,
  };
  roundRecord.updated_at = completedAt;
  session.status = {
    ...session.status,
    active_actor: error ? actor : null,
    active_phase: error ? phase : null,
    error: error
      ? {
          message: error.message || String(error),
          at: completedAt,
        }
      : null,
  };
}

function noteRoundEvidence(session, { round, evidence_added }) {
  const roundRecord = ensureOpenRound(session, round);
  invariant(Array.isArray(evidence_added), "evidence_added must be an array");
  const seen = new Set(roundRecord.evidence_added);
  for (const evidenceId of evidence_added) {
    invariant(typeof evidenceId === "string" && evidenceId.trim().length > 0, "evidence_added[] must be non-empty strings");
    if (!seen.has(evidenceId)) {
      roundRecord.evidence_added.push(evidenceId);
      seen.add(evidenceId);
    }
  }
  roundRecord.updated_at = nowIso();
}

function registerProposal(session, { round, proposal }) {
  validateProposal(proposal);
  const roundRecord = ensureOpenRound(session, round);
  roundRecord.proposal = proposal;
  roundRecord.updated_at = nowIso();
}

function appendRoundFindings(session, { round, findings }) {
  const roundRecord = ensureOpenRound(session, round);
  invariant(Array.isArray(findings), "findings must be an array");
  const evidence = evidenceIndex(session);
  const seenFindingIds = new Set(currentRoundFindings(roundRecord).map((item) => item.id));
  const nextFindings = findings.map((finding, index) => {
    const next = { ...finding };
    validateFinding(next, evidence, `open_round.findings_against_proposal[${index}]`);
    invariant(!seenFindingIds.has(next.id), `duplicate finding id ${next.id}`);
    seenFindingIds.add(next.id);
    return next;
  });
  roundRecord.findings_against_proposal.push(...nextFindings);
  refreshOpenRoundSummary(session);
}

function registerVerdict(session, { round, verdict }) {
  const roundRecord = ensureOpenRound(session, round);
  if (verdict === null) {
    roundRecord.verdict = null;
    roundRecord.updated_at = nowIso();
    return;
  }
  invariant(verdict && typeof verdict === "object", "verdict must be an object");
  invariant(Array.isArray(verdict.decisions), "verdict.decisions must be an array");
  if (verdict.revised_proposal !== undefined && verdict.revised_proposal !== null) {
    validateProposal(verdict.revised_proposal, "verdict.revised_proposal");
  }
  const findings = currentRoundFindings(roundRecord);
  const findingsById = findingIndex(findings);
  const evidence = evidenceIndex(session);
  invariant(verdict.decisions.length === findings.length, "verdict must contain exactly one decision for every finding");
  const seenDecisionIds = new Set();
  verdict.decisions.forEach((decision, index) => {
    validateDecision(decision, findingsById, evidence, `open_round.verdict.decisions[${index}]`);
    invariant(!seenDecisionIds.has(decision.finding_id), `duplicate decision for finding ${decision.finding_id}`);
    seenDecisionIds.add(decision.finding_id);
  });
  roundRecord.verdict = verdict;
  roundRecord.updated_at = nowIso();
}

function markSessionFailed(session, { round, actor, phase, error }) {
  const message = error?.message || String(error);
  session.status = {
    ...session.status,
    round: round || session.open_round?.index || session.status.round,
    state: "failed",
    active_actor: actor || session.status.active_actor,
    active_phase: phase || session.status.active_phase,
    error: {
      message,
      at: nowIso(),
    },
  };
  if (session.open_round) {
    session.open_round.error = {
      message,
      actor: actor || null,
      phase: phase || null,
      at: nowIso(),
    };
    session.open_round.updated_at = nowIso();
  }
}

function applyRound(session) {
  const roundRecord = session.open_round;
  invariant(roundRecord, "session has no open round to apply");
  validateProposal(roundRecord.proposal);
  invariant(Array.isArray(roundRecord.evidence_added), "open_round.evidence_added must be an array");

  const findings = currentRoundFindings(roundRecord).map((finding) => ({ ...finding }));
  const counts = findingCounts(findings);
  const material = findings.filter(isMaterialFinding);
  const verdict = roundRecord.verdict;

  if (verdict) {
    invariant(verdict && typeof verdict === "object", "verdict must be an object");
    invariant(Array.isArray(verdict.decisions), "verdict.decisions must be an array");
  } else {
    invariant(material.length === 0, "material findings require a verdict before the round can close");
  }

  const closedRound = {
    index: roundRecord.index,
    proposal: roundRecord.proposal,
    evidence_added: [...roundRecord.evidence_added],
    findings_against_proposal: findings,
    review_summary: counts,
    verdict,
    phase_history: [...roundRecord.phase_history],
    created_at: roundRecord.created_at,
    updated_at: nowIso(),
  };

  session.rounds.push(closedRound);
  session.open_round = null;
  session.adjudicated_proposal = verdict?.revised_proposal || roundRecord.proposal;
  session.status = {
    ...session.status,
    round: closedRound.index,
    converged: counts.material === 0,
    unresolved_high: counts.high,
    unresolved_medium: counts.medium,
    state: counts.material === 0 ? "converged" : "needs_revision",
    active_actor: null,
    active_phase: null,
    error: null,
  };

  return closedRound;
}

module.exports = {
  appendEvidence,
  applyRound,
  appendRoundFindings,
  completePhase,
  createSession,
  markSessionFailed,
  noteRoundEvidence,
  recordPhaseStart,
  registerProposal,
  registerVerdict,
};

File: internal/rtk/kernel.go
package rtk

import (
	"fmt"
	"regexp"
	"time"
)

var sessionIDPattern = regexp.MustCompile(`^[A-Za-z0-9._-]+$`)

func nowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
}

func makeEvidenceID(session *Session) string {
	return fmt.Sprintf("E%d", len(session.Evidence)+1)
}

func evidenceIndex(session *Session) map[string]Evidence {
	index := make(map[string]Evidence, len(session.Evidence))
	for _, item := range session.Evidence {
		index[item.ID] = item
	}
	return index
}

func findingIndex(findings []Finding) map[string]Finding {
	index := make(map[string]Finding, len(findings))
	for _, item := range findings {
		index[item.ID] = item
	}
	return index
}

func ensureOpenRound(session *Session, round int) (*RoundRecord, error) {
	if err := invariant(round > 0, "open round index must be a positive integer"); err != nil {
		return nil, err
	}
	if session.OpenRound == nil {
		session.OpenRound = &RoundRecord{
			Index:                   round,
			Proposal:                nil,
			EvidenceAdded:           []string{},
			FindingsAgainstProposal: []Finding{},
			ReviewSummary:           Counts{},
			Verdict:                 nil,
			PhaseHistory:            []PhaseRecord{},
			CreatedAt:               nowISO(),
			UpdatedAt:               nowISO(),
			Error:                   nil,
		}
	}
	if session.OpenRound.Index != round {
		return nil, fmt.Errorf("session already has open round %d", session.OpenRound.Index)
	}
	return session.OpenRound, nil
}

func refreshOpenRoundSummary(session *Session) {
	if session.OpenRound == nil {
		return
	}
	counts := FindingCounts(session.OpenRound.CurrentFindings())
	session.OpenRound.ReviewSummary = counts
	session.OpenRound.UpdatedAt = nowISO()
	session.Status.Round = session.OpenRound.Index
	session.Status.UnresolvedHigh = counts.High
	session.Status.UnresolvedMedium = counts.Medium
}

func roundPhaseIndex(roundRecord *RoundRecord, actor, phase, status string) int {
	for index := len(roundRecord.PhaseHistory) - 1; index >= 0; index-- {
		item := roundRecord.PhaseHistory[index]
		if item.Actor == actor && item.Phase == phase && item.Status == status {
			return index
		}
	}
	return -1
}

func NewSession(id, topic, chair string, critics []string, maxRounds *int, adapter string) (*Session, error) {
	if err := validateString(id, "session id"); err != nil {
		return nil, err
	}
	if !sessionIDPattern.MatchString(id) {
		return nil, fmt.Errorf("session id must match [A-Za-z0-9._-]+")
	}
	if err := validateString(topic, "topic"); err != nil {
		return nil, err
	}
	if err := validateString(chair, "chair"); err != nil {
		return nil, err
	}
	if maxRounds != nil {
		if err := invariant(*maxRounds > 0, "max_rounds must be a positive integer"); err != nil {
			return nil, err
		}
	}
	if err := validateString(adapter, "adapter"); err != nil {
		return nil, err
	}
	return &Session{
		Version:             1,
		ID:                  id,
		Topic:               topic,
		CreatedAt:           nowISO(),
		Chair:               chair,
		Critics:             append([]string{}, critics...),
		MaxRounds:           cloneIntPtr(maxRounds),
		Adapter:             adapter,
		Evidence:            []Evidence{},
		Rounds:              []RoundRecord{},
		OpenRound:           nil,
		AdjudicatedProposal: nil,
		Status: Status{
			Round:            0,
			Converged:        false,
			UnresolvedHigh:   0,
			UnresolvedMedium: 0,
			State:            "initialized",
			ActiveActor:      nil,
			ActivePhase:      nil,
			Error:            nil,
		},
	}, nil
}

type EvidenceInput struct {
	Key       string `json:"key,omitempty"`
	Source    string `json:"source"`
	Kind      string `json:"kind"`
	Statement string `json:"statement"`
	Excerpt   string `json:"excerpt"`
}

func AppendEvidence(session *Session, items []EvidenceInput, collectedBy, phase string, round int) ([]Evidence, error) {
	if err := validateString(collectedBy, "collectedBy"); err != nil {
		return nil, err
	}
	if err := validateString(phase, "phase"); err != nil {
		return nil, err
	}
	if err := invariant(round >= 0, "round must be a non-negative integer"); err != nil {
		return nil, err
	}
	added := make([]Evidence, 0, len(items))
	for _, item := range items {
		evidence := Evidence{
			ID:          makeEvidenceID(session),
			Source:      item.Source,
			Kind:        item.Kind,
			Phase:       phase,
			Statement:   item.Statement,
			Excerpt:     item.Excerpt,
			CollectedBy: collectedBy,
			Round:       round,
			CreatedAt:   nowISO(),
		}
		if err := ValidateEvidence(&evidence, "evidence"); err != nil {
			return nil, err
		}
		session.Evidence = append(session.Evidence, evidence)
		added = append(added, evidence)
	}
	return added, nil
}

func startRound(session *Session, round int) (*RoundRecord, error) {
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return nil, err
	}
	session.Status.Round = round
	session.Status.Converged = false
	session.Status.UnresolvedHigh = roundRecord.ReviewSummary.High
	session.Status.UnresolvedMedium = roundRecord.ReviewSummary.Medium
	session.Status.State = "running"
	session.Status.ActiveActor = nil
	session.Status.ActivePhase = nil
	session.Status.Error = nil
	roundRecord.UpdatedAt = nowISO()
	return roundRecord, nil
}

func RecordPhaseStart(session *Session, round int, actor, phase string, inputSummary map[string]any) error {
	roundRecord, err := startRound(session, round)
	if err != nil {
		return err
	}
	roundRecord.PhaseHistory = append(roundRecord.PhaseHistory, PhaseRecord{
		Actor:         actor,
		Phase:         phase,
		Status:        "running",
		InputSummary:  inputSummary,
		OutputSummary: nil,
		Artifact:      nil,
		Usage:         nil,
		StartedAt:     nowISO(),
		CompletedAt:   nil,
		DurationMS:    nil,
		Error:         nil,
	})
	roundRecord.UpdatedAt = nowISO()
	session.Status.ActiveActor = stringPtr(actor)
	session.Status.ActivePhase = stringPtr(phase)
	session.Status.Error = nil
	return nil
}

func CompletePhase(session *Session, round int, actor, phase string, outputSummary map[string]any, artifact map[string]any, usage *PhaseUsage, durationMS int64, phaseErr error) error {
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return err
	}
	index := roundPhaseIndex(roundRecord, actor, phase, "running")
	if index < 0 {
		return fmt.Errorf("no running phase record for %s/%s in round %d", actor, phase, round)
	}
	status := "succeeded"
	var errorInfo *SessionError
	if phaseErr != nil {
		status = "failed"
		errorInfo = &SessionError{
			Message: phaseErr.Error(),
			At:      nowISO(),
		}
	}
	completedAt := nowISO()
	roundRecord.PhaseHistory[index] = PhaseRecord{
		Actor:         roundRecord.PhaseHistory[index].Actor,
		Phase:         roundRecord.PhaseHistory[index].Phase,
		Status:        status,
		InputSummary:  roundRecord.PhaseHistory[index].InputSummary,
		OutputSummary: outputSummary,
		Artifact:      artifact,
		Usage:         normalizePhaseUsage(usage),
		StartedAt:     roundRecord.PhaseHistory[index].StartedAt,
		CompletedAt:   &completedAt,
		DurationMS:    int64Ptr(durationMS),
		Error:         errorInfo,
	}
	roundRecord.UpdatedAt = completedAt
	if phaseErr != nil {
		session.Status.ActiveActor = stringPtr(actor)
		session.Status.ActivePhase = stringPtr(phase)
		session.Status.Error = &SessionError{Message: phaseErr.Error(), At: completedAt}
	} else {
		session.Status.ActiveActor = nil
		session.Status.ActivePhase = nil
		session.Status.Error = nil
	}
	return nil
}

func NoteRoundEvidence(session *Session, round int, evidenceAdded []string) error {
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return err
	}
	seen := make(map[string]bool, len(roundRecord.EvidenceAdded))
	for _, item := range roundRecord.EvidenceAdded {
		seen[item] = true
	}
	for _, evidenceID := range evidenceAdded {
		if err := validateString(evidenceID, "evidence_added[]"); err != nil {
			return err
		}
		if !seen[evidenceID] {
			roundRecord.EvidenceAdded = append(roundRecord.EvidenceAdded, evidenceID)
			seen[evidenceID] = true
		}
	}
	roundRecord.UpdatedAt = nowISO()
	return nil
}

func RegisterProposal(session *Session, round int, proposal *Proposal) error {
	if err := ValidateProposal(proposal, "proposal"); err != nil {
		return err
	}
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return err
	}
	roundRecord.Proposal = proposal
	roundRecord.UpdatedAt = nowISO()
	return nil
}

func AppendRoundFindings(session *Session, round int, findings []Finding) error {
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return err
	}
	evidence := evidenceIndex(session)
	seen := make(map[string]bool, len(roundRecord.CurrentFindings()))
	for _, item := range roundRecord.CurrentFindings() {
		seen[item.ID] = true
	}
	for index, finding := range findings {
		copy := finding
		if err := ValidateFinding(&copy, evidence, fmt.Sprintf("open_round.findings_against_proposal[%d]", index)); err != nil {
			return err
		}
		if seen[copy.ID] {
			return fmt.Errorf("duplicate finding id %s", copy.ID)
		}
		seen[copy.ID] = true
		roundRecord.FindingsAgainstProposal = append(roundRecord.FindingsAgainstProposal, copy)
	}
	refreshOpenRoundSummary(session)
	return nil
}

func RegisterVerdict(session *Session, round int, verdict *Verdict) error {
	roundRecord, err := ensureOpenRound(session, round)
	if err != nil {
		return err
	}
	if verdict == nil {
		roundRecord.Verdict = nil
		roundRecord.UpdatedAt = nowISO()
		return nil
	}
	if verdict.RevisedProposal != nil {
		if err := ValidateProposal(verdict.RevisedProposal, "verdict.revised_proposal"); err != nil {
			return err
		}
	}
	findings := roundRecord.CurrentFindings()
	findingByID := findingIndex(findings)
	evidence := evidenceIndex(session)
	if len(verdict.Decisions) != len(findings) {
		return fmt.Errorf("verdict must contain exactly one decision for every finding")
	}
	seen := map[string]bool{}
	for index, decision := range verdict.Decisions {
		copy := decision
		if err := ValidateDecision(&copy, findingByID, evidence, fmt.Sprintf("open_round.verdict.decisions[%d]", index)); err != nil {
			return err
		}
		if seen[copy.FindingID] {
			return fmt.Errorf("duplicate decision for finding %s", copy.FindingID)
		}
		seen[copy.FindingID] = true
	}
	roundRecord.Verdict = verdict
	roundRecord.UpdatedAt = nowISO()
	return nil
}

func MarkSessionFailed(session *Session, round int, actor, phase string, failure error) {
	message := "unknown error"
	if failure != nil {
		message = failure.Error()
	}
	at := nowISO()
	session.Status.Round = round
	if session.Status.Round == 0 && session.OpenRound != nil {
		session.Status.Round = session.OpenRound.Index
	}
	session.Status.State = "failed"
	if actor != "" {
		session.Status.ActiveActor = stringPtr(actor)
	}
	if phase != "" {
		session.Status.ActivePhase = stringPtr(phase)
	}
	session.Status.Error = &SessionError{Message: message, At: at}
	if session.OpenRound != nil {
		session.OpenRound.Error = &SessionError{
			Message: message,
			Actor:   stringPtr(actor),
			Phase:   stringPtr(phase),
			At:      at,
		}
		session.OpenRound.UpdatedAt = at
	}
}

func ApplyRound(session *Session) (*RoundRecord, error) {
	if session.OpenRound == nil {
		return nil, fmt.Errorf("session has no open round to apply")
	}
	roundRecord := session.OpenRound
	if err := ValidateProposal(roundRecord.Proposal, "open_round.proposal"); err != nil {
		return nil, err
	}
	findings := append([]Finding{}, roundRecord.CurrentFindings()...)
	counts := FindingCounts(findings)
	material := make([]Finding, 0, counts.Material)
	for _, finding := range findings {
		if IsMaterialFinding(finding) {
			material = append(material, finding)
		}
	}
	if roundRecord.Verdict == nil && len(material) > 0 {
		return nil, fmt.Errorf("material findings require a verdict before the round can close")
	}
	closed := RoundRecord{
		Index:                   roundRecord.Index,
		Proposal:                roundRecord.Proposal,
		EvidenceAdded:           append([]string{}, roundRecord.EvidenceAdded...),
		FindingsAgainstProposal: findings,
		ReviewSummary:           counts,
		Verdict:                 roundRecord.Verdict,
		PhaseHistory:            append([]PhaseRecord{}, roundRecord.PhaseHistory...),
		CreatedAt:               roundRecord.CreatedAt,
		UpdatedAt:               nowISO(),
	}
	session.Rounds = append(session.Rounds, closed)
	session.OpenRound = nil
	if closed.Verdict != nil && closed.Verdict.RevisedProposal != nil {
		session.AdjudicatedProposal = closed.Verdict.RevisedProposal
	} else {
		session.AdjudicatedProposal = closed.Proposal
	}
	session.Status.Round = closed.Index
	session.Status.Converged = counts.Material == 0
	session.Status.UnresolvedHigh = counts.High
	session.Status.UnresolvedMedium = counts.Medium
	if counts.Material == 0 {
		session.Status.State = "converged"
	} else {
		session.Status.State = "needs_revision"
	}
	session.Status.ActiveActor = nil
	session.Status.ActivePhase = nil
	session.Status.Error = nil
	return &closed, nil
}
